docs/source/tutorial_vectorarray.md

```diff
@@ -0,0 +1,687 @@
+V.impl._array
+```
+
+As NumPy arrays store their data consecutively in memory, this can significantly improve
+performance of iterative algorithms where new vectors are repeatedly appended to a given array.
+However, this is only useful for implementations which do store the data consecutively in memory.
+
+We will now look at another type of {{ VectorArray }}, which uses Python lists of one-dimensional
+NumPy arrays to store the data as a list of vectors:
+
+```{code-cell}
+from pymor.vectorarrays.list import NumpyListVectorSpace
+space = NumpyListVectorSpace(4)
+W = space.random(3)
+```
+
+`W` behaves as any other array.
+It also implements `to_numpy` so that we can look at the data directly:
+
+```{code-cell}
+W.to_numpy()
+```
+
+However, the implementation class now holds a list of vector objects:
+
+```{code-cell}
+print(type(W))
+print(type(W.impl))
+print(W.impl._list)
+print(type(W.impl._list))
+```
+
+Each of these {class}`~pymor.vectorarrays.list.NumpyVector` objects now stores one of the vectors
+in the array:
+
+```{code-cell}
+W.impl._list[1]._array
+```
+
+Ending this tutorial, we want to take a brief look how one of the interface methods is actually
+implemented by the different array types.
+We choose {meth}`~pymor.vectorarrays.interface.VectorArray.pairwise_inner` as the implementation is
+particularly simple.
+Here is the implementation for {{ NumpyVectorArray }}:
+
+```{code-cell}
+from pymor.tools.formatsrc import print_source
+print_source(U.impl.pairwise_inner)
+```
+
+We see the actual call to NumPy that does all computations in the last line.
+The implementation has to take into account that one of the arrays might be a view, in which case the
+selected indices are given by `ind`/`oind`.
+Also, `reserve` might have been used when the array was created, making `_array` larger than it
+actually is.
+The true length is stored in the `_len` attribute, so that the first `_len` vectors from the array
+are selected.
+
+Here is the code for the `list`-based array:
+
+```{code-cell}
+print_source(W.impl.pairwise_inner)
+```
+
+`W.impl` is a generic {class}`~pymor.vectorarrays.list.ListVectorArrayImpl` object, which provides a
+common implementation for all {{ VectorArrays }} based on Python lists.
+The implementation just loops over all pairs `a, b` of vector objects in both arrays and calls `a`'s
+{meth}`~pymor.vectorarrays.interface.VectorArray.inner` method, which computes the actual inner product.
+In case of {class}`~pymor.vectorarrays.list.NumpyVector`, the code looks like this:
+
+```{code-cell}
+print_source(W.impl._list[0].inner)
+```
+
+As most PDE solver libraries only know of single vectors and have no notion of an array of vectors,
+{{ VectorArrays }} for external solvers are typically based on {{ ListVectorArray }} /
+{class}`~pymor.vectorarrays.list.ListVectorArrayImpl`.
+In that case, only the interface methods for a single {class}`~pymor.vectorarrays.list.Vector` object
+have to be implemented.
```

Comment:
I have added 
```
Note that for NumPy-based data, {{ NumpyVectorArray }} is the default {{ VectorArray }} type used throughout
pyMOR, as it can benefit from vectorized operations on the underlying NumPy array.
{class}`~pymor.vectorarrays.list.NumpyListVectorSpace` is only implemented in pyMOR for testing purposes.
```